src/rust/engine/fs/src/snapshot.rs

```diff
@@ -119,22 +116,69 @@ least one path component, but at least one did not"
         directory.set_directories(protobuf::RepeatedField::from_vec(dirs));
         directory.set_files(protobuf::RepeatedField::from_vec(files));
         let digest = store.record_directory(&directory)?;
-        Ok(Snapshot {
-          fingerprint: digest.0,
-          digest: Some(digest),
-          path_stats: path_stats,
-        })
+        Ok(Snapshot { digest, path_stats })
       })
       .to_boxed()
   }
+
+  // TODO: How much should we be aiming to parallelise this?
+  // The old method returned a Future which did all of the work in one work-unit, on the PosixFS
+  // CpuPool. Is that sufficient? Or should we be kicking off each LMDB read in parallel on the
+  // pool? I can start investigating the performance of parallel LMDB reads on APFS, but I wanted to
+  // check whether it was even worth thinking about before doing so.
+  //
+  // (This is currently not done in a pool at all because I want to resolve the TODO on PosixFS's
+  // pool() method before doing so).
+  pub fn contents(self, store: Arc<Store>) -> BoxFuture<Vec<FileContent>, String> {
+    futures::future::done(Snapshot::contents_for_directory_sync(
+      &self.digest.0,
+      store,
+      PathBuf::from(""),
+    )).to_boxed()
+  }
+
+  fn contents_for_directory_sync(
+    fingerprint: &Fingerprint,
+    store: Arc<Store>,
+    path_so_far: PathBuf,
+  ) -> Result<Vec<FileContent>, String> {
+    let directory = store.load_directory_proto(fingerprint)?.ok_or_else(|| {
+      format!("Could not find snapshot {}", fingerprint)
+    })?;
+
+    let mut contents = Vec::new();
+
+    for file_node in directory.get_files() {
+      contents.push(FileContent {
+        path: path_so_far.join(file_node.get_name()),
+        content: store
+          .load_file_bytes(&Fingerprint::from_hex_string(
+            file_node.get_digest().get_hash(),
+          ).unwrap())?
+          .unwrap(),
+      });
+    }
+
+    for dir_node in directory.get_directories() {
+      contents.extend(
+        Snapshot::contents_for_directory_sync(
+          &Fingerprint::from_hex_string(dir_node.get_digest().get_hash()).unwrap(),
+          store.clone(),
+          path_so_far.join(dir_node.get_name()),
+        ).unwrap(),
+      );
+    }
+
+    contents.sort_by(|a, b| a.path.cmp(&b.path));
```

Comment:
This will have the effect of sorting too many times due to the recursion.

If you guarantee/assert that the `Directory` proto has sorted file and directory lists (...maybe we should push this guarantee upstream...?), this gets a bit easier. Assert sorted, and then either:
1. Always render the files in a directory before the directories in a directory.
2. Execute a linear-time merge of the sorted `DirNode` and `FileNode` lists.